### DISCUSSION TITLE
MyPub progress diary

### DISCUSSION THREAD

**Discussion by WesSouza** [WesSouza/MyPub#1] [Diary]
## 💾 10 PRINT "Hellorld" 20 GOTO 10

Hello. This is my public diary for progress of the project. I could use "Projects" but who likes Jira anyway. It worked for me with [Mastodon Flock](https://mastodon-flock.vercel.app), so I'm doing it again, but this time on the open.

### 💡 Idea & Architecture

This will be a set of libraries that lets people host their own ActivityPub instance on an existing website. It'll be built as JS functions combined with adapters for different servers, like Express.

The end result will provide two layers of endpoints: an ActivityPub layer to send and receive content and user interaction, as well as a Mastodon-compatible API to allow any application to interact with the instance, so you can use it without going to the database.

Storage and adapters will be interchangeable, allowing multiple use cases under the same principle of a single-user ActivityPub instance.

A reference front-end might be also provided, but only to allow rendering of user content, given all login and interactions are handled on an app.

### 🚧 In Development

- [ ] (current) Send and receive activities:
  - [ ] `Send:Create(Note)`
  - [ ] `Send:Update(Note)`
  - [ ] `Send:Delete(Note)`
  - [ ] `Receive:Create(Note)`
  - [ ] `Receive:Update(Note)`
  - [ ] `Receive:Delete(Note)`
  - [ ] `Send:Like(Note)`
  - [ ] `Send:Undo(Like)`
  - [ ] `Send:Announce(Note)`
  - [ ] `Send:Undo(Announce)`
  - [ ] `Receive:Like(Note)`
  - [ ] `Receive:Undo(Like)`
  - [ ] `Receive:Announce(Note)`
  - [ ] `Receive:Undo(Announce)`
- [ ] Observe the behavior of the following activities from Mastodon:
  - [ ] `Ignore`
  - [ ] `Block`
- [ ] Mastodon API
  - [ ] Investigate basic needed APIs (onboarding + first screen on an app)
  - [ ] Implement the APIs above
- [ ] Adapters
  - [ ] Express
  - [ ] Next.js
  - [ ] Astro
  - [ ] 11y
  - [ ] Deno
- [x] Reading about ActivityPub
  - [x] [the honk guy](https://flak.tedunangst.com/post/ActivityPub-as-it-has-been-understood): what a shit show AP is
  - [x] https://socialhub.activitypub.rocks/pub/guide-for-new-activitypub-implementers
  - [x] https://tinysubversions.com/notes/reading-activitypub/
  - [x] https://www.w3.org/TR/activitypub/
  - [x] https://www.w3.org/TR/activitystreams-core/
  - [x] https://www.w3.org/TR/activitystreams-vocabulary/
- [x] Try: Basic ActivityPub types via Zod, given I will need to validate a shit ton of API responses
- [x] Try: Sample implementation of an ActivityPub instance using TS, Fastify, MongoDB to prove the point
  - [x] `Send:Follow`, `Receive:Accept(Follow)`
  - [x] `Receive:Follow`, `Send:Accept(Follow)`
  - [x] `Receive:Undo(Follow)`
  - [x] `Send:Undo(Follow)`
- [x] Install [Honk](https://humungus.tedunangst.com/r/honk) somewhere cheap, ideally free

(and 1000s more)

### 🚨 Problems Found So Far

- **No Likes and Shares information** although ActivityPub allows objects to have `likes` and `shares` collections, it seems Mastodon does not send them, and there's no mechanism to synchronize that. This is very bad for any federated timeline, which only knows about interactions that happened on itself.
- **Delivery will be challenging** on serverless, due to the fact activities must be individually shared among all your followers by yourself. Is there no such thing as ActivityRelay?
- **Follower collection synchronization** exists, it's [minimally documented](https://docs.joinmastodon.org/spec/activitypub/#follower-synchronization-mechanism), and I have no idea when it is applicable or how it actually looks like. Do I really need to learn Ruby 😑

### 🫂 Helping

You can help me by starring this project, providing comments and insights, buying me a coffee or talking me out of it by showing how someone else already did it and I'll be wasting my time.

For now, given the structure and plan is all in my head, I don't want to manage user stories and tickets. Once the project is on an alpha state, I'd love external code contributions.

And, of course, user testing. Once this has a relatively stable version, use it, break it, complain about it.

**Comment by weex** [comment]
Starred. My main insight is if you're looking to build a community around this to choose a copyleft license. Non-copyleft is a non-starter for because it enables proprietary competition that can always be a step ahead. I'm not sure how active it is now but my first months getting to know the fediverse we're with a group who wanted to build AP into a TS couchsurfing app (Trustroots).

**Reply by WesSouza** [reply]
@weex can you point me to more information about common copyleft licenses? How would that be different than the MIT license I usually choose?

I am all in favor of free software licensing, as long as one of its tenets is to requiring attribution.

(by the way, I'm going to delete your comment here soon and lock the Diary so that I can use the comments as the actual diary updates, feel free to [recreate this discussion as its own post](https://github.com/WesSouza/MyPub/discussions/new?category=General&title=Copyleft%20License))

**Reply by PaoloM** [reply]
Please keep it MIT.
